mpi4jax/collective_ops/allreduce.py

```diff
@@ -51,16 +51,25 @@ def Allreduce(x, op, comm=_MPI.COMM_WORLD, token=None):
         new_token: a new, modified token, that depends on this operation.
             This result can be ignored if result forces a data dependency.
     """
+
+    # The extra argument _transpose is an implementation detail. It is used to
+    # keep track of whever we are computing the forward or transposition of
+    # Allreduce, because we are 'cheating' and not performing MPI operations
+    # on the tranposed (even though, in principle, we should) in order
+    # to be more efficient.
```

Comment:
Which operations are meant by `even though, in principle, we should` ?

Reply:
I mean, in theory you should collect the output of all ranks on every rank (Which is a MPI.Gather operation), and then use all that information to compute the transpose pass.

But as all ranks have the same output, we don't need to do it.

Think for example at implementing the transpose for MPI.Reduce, which gives the same output of Allreduce but only on one rank. In principle you'd need to Send the output  of MPI.Reduce present on the master rank to all other ranks